### ISSUE TITLE
Adapt ART to other types of models than neural networks

### ISSUE THREAD

**Issue by ririnicolae** [IBM/adversarial-robustness-toolbox#50]
This mostly implies changing the way different backend implementations handle model outputs. In order to be effective, most attacks need to be applied at logits level. Some of the new models types ART will support do not have logits. The following describes the proposed solution for this problem:

- All backend `__init__` functions would have a flag specifying if the output of the given model represents logits. When such a model is created, attacks are applied on the regular model output, as was done previously.
- If the outputs are not logits, we can check if the sum of the outputs on a fake input is 1. In that case, we have detected probabilities as outputs and will try to estimate the logits from them. These would not be the real logits of the model, but would perform numerical scaling that makes attacks efficient.
- If the outputs are not logits and don't sum to 1 either, this might be a different type of model, and attacks will be applied directly on its outputs.
- Another flag could indicate that the outputs of the model are probabilities, but the user explicitly requests that the previous layer (logits) should be used for attacks. This option is specific to the frameworks that support the notion of layer. Moreover, finding the correct logits would not be guaranteed.
- The loss function for a given model will also be part of the input provided by the user. The loss will be applied directly on the output of the model.

In practice, the implications are that predictions / gradient computations will not be available at both logits and probabilities level, as it is currently the case. This would attract a change in the `Classifier` API for the respective functions.

**Comment by minhitbk** [comment]
I will help on Pytorch and TF classifiers.

**Comment by ririnicolae** [comment]
The branch for these features is `other_models`.

**Comment by ririnicolae** [comment]
Closed in #119.
